src/emulator/host/src/config.cpp

```diff
@@ -27,14 +27,126 @@
 #include <boost/program_options.hpp>
 #include <spdlog/spdlog.h>
 
+#include <yaml-cpp/yaml.h>
+
 #include <exception>
+#include <fstream>
 #include <iostream>
 
 namespace po = boost::program_options;
 
 namespace config {
 
+template <typename T, typename Q = T>
+void get_yaml_value(YAML::Node &config_node, const char *key, T *target_val, Q default_val) {
+    try {
+        *target_val = config_node[key].as<Q>();
+    } catch (YAML::Exception &exception) {
+        *target_val = std::move(default_val);
+    }
+}
+
+template <typename T>
+void get_yaml_value_optional(YAML::Node &config_node, const char *key, boost::optional<T> *target_val, T default_val) {
+    try {
+        *target_val = config_node[key].as<T>();
+    } catch (YAML::Exception &exception) {
+        *target_val = std::move(default_val);
+    }
+}
+
+template <typename T>
+void get_yaml_value_optional(YAML::Node &config_node, const char *key, boost::optional<T> *target_val) {
+    try {
+        *target_val = config_node[key].as<T>();
+    } catch (YAML::Exception &exception) {
+        *target_val = boost::none;
+    }
+}
+
+template <typename T>
+void config_file_emit_single(YAML::Emitter &emitter, const char *name, T &val) {
+    emitter << YAML::Key << name << YAML::Value << val;
+}
+
+template <typename T>
+void config_file_emit_optional_single(YAML::Emitter &emitter, const char *name, boost::optional<T> &val) {
+    if (val) {
+        emitter << YAML::Key << name << YAML::Value << val.value();
```

Comment:
could re-use `config_file_emit_single` here, if you want

Reply:
YAML can't serialize boost::optional though

Reply:
I thought you could call with `val.value()`.

Reply:
Ah, I taken your idea wrong. it's good